Docusaurus v2.0 update

prakharsr opened the pull request:
In reference to issue https://github.com/accordproject/techdocs/issues/160, please see my comment at https://github.com/accordproject/techdocs/issues/160#issuecomment-725095044.

mttrbrts commented:
Thanks for your hard work @prakharsr.

I suggest that we create a development branch ready for v2 release so that we can support multiple contributions until we're ready to make it the `master` branch. I've created the branch (https://github.com/accordproject/techdocs/tree/docusaurus-v2), so you should retarget your PR to that.

That will also mean that you can promote your `v2-website` directory to the top-level and we can remove the old content.

We can then create separate issues for specific issues and ask others for help.

Finally, are you able to fix the DCO sign-off, please? Let me know if you need help with this.

prakharsr commented:
Hi, @mttrbrts . Thanks for helping, I have rebased my branch from master to docusaurus-v2 and fixed the DCO signoff issue.
The features/ components which are incomplete and need fixing are:
1.     Fixing the index page and footer and make it look the same as in v1 (I am facing hurdles here as I don't know much react).
2.     Fixing background colors and snippets color highlighting.
3.     Migrating syntax highlighting to Prism (not yet completed).
4.     I was unable to build the website due to (undefined) Error: Minified React error #152, leading to unable deploy the static website at github pages. A development version is currently being deployed by me at https://accordproject-znyzg7zhdq-uc.a.run.app/ where a preview of the migrated site can be seen.

mttrbrts commented:
Thanks @prakharsr.

There's a lot of files here for us to review, so please bear with us while we do that.

There are two things that would be really helpful for the reviewers:

1. Please delete the existing `website` folder, then rename `v2-website` to `website`. We will use branches to manage concurrent versions instead. This will mean that you can remove all changes to docs and config that just deal with directory changes.
2. Please add `.docusarus` to `.gitignore` and delete the existing `.docusaurus1 directory. This content is dynamic and doesn't need to be checked into the repo.

irmerk commented:
Looks like the error is something with webpack?

```sh
11:33:58 PM: > docusaurus build
11:33:58 PM: /opt/build/repo/website/node_modules/copy-webpack-plugin/node_modules/p-limit/index.js:30
11:33:58 PM: 		} catch {}
11:33:58 PM: 		        ^
11:33:58 PM: SyntaxError: Unexpected token {
11:33:58 PM:     at createScript (vm.js:80:10)
11:33:58 PM:     at Object.runInThisContext (vm.js:139:10)
11:33:58 PM:     at Module._compile (module.js:617:28)
11:33:58 PM:     at Object.Module._extensions..js (module.js:664:10)
11:33:58 PM:     at Module.load (module.js:566:32)
11:33:58 PM:     at tryModuleLoad (module.js:506:12)
11:33:58 PM:     at Function.Module._load (module.js:498:3)
11:33:58 PM:     at Module.require (module.js:597:17)
11:33:58 PM:     at require (internal/module.js:11:18)
11:33:58 PM:     at Object.<anonymous> (/opt/build/repo/website/node_modules/copy-webpack-plugin/dist/index.js:18:38)
11:33:58 PM: npm
11:33:58 PM:  ERR! code
11:33:58 PM:  ELIFECYCLE
11:33:58 PM: npm
11:33:58 PM: ERR! errno 1
11:33:58 PM: npm
11:33:58 PM:  ERR!
11:33:58 PM:  @ build: `docusaurus build`
11:33:58 PM: npm
11:33:58 PM:  ERR! Exit status 1
11:33:58 PM: npm
11:33:58 PM:  ERR!
11:33:58 PM: npm
11:33:58 PM:  ERR!
11:33:58 PM:  Failed at the @ build script.
11:33:58 PM: npm
11:33:58 PM:  ERR! This is probably not a problem with npm. There is likely additional logging output above.
11:33:58 PM: npm
11:33:58 PM: ERR! A complete log of this run can be found in:
11:33:58 PM: npm
11:33:58 PM: ERR!     /opt/buildhome/.npm/_logs/2020-12-02T04_33_58_583Z-debug.log
11:33:58 PM: Skipping functions preparation step: no functions directory set
```

prakharsr commented:
Yeah, probably, though I got a different error message.`^C22:17:36|prakhar@hp-ay503tx|website$ npm run build

> @ build /home/prakhar/projects/techdocs_og/techdocs/website
> docusaurus build

Creating an optimized production build...
The docs plugin option homePageId=accordproject is deprecated. To make a doc the "home", prefer frontmatter: "slug: /"

✔ Client
  Compiled successfully in 3.74m

✖ Server
  Compiled with some errors in 4.19m



Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
(undefined) Error: Minified React error #152; visit https://reactjs.org/docs/error-decoder.html?invariant=152&args[]=code for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Za (main:69598:465)
    at d (main:69600:55)
    at $a (main:69602:16)
    at a.b.render (main:69607:476)
    at a.b.read (main:69607:18)
    at Object.renderToString (main:69617:364)
    at serverEntry_render (main:96653:664)
Error: Failed to compile with errors.
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/@docusaurus/core/lib/webpack/utils.js:163:24
    at finalCallback (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/MultiCompiler.js:254:12)
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/MultiCompiler.js:277:6
    at done (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/neo-async/async.js:2931:13)
    at runCompilers (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/MultiCompiler.js:181:48)
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/MultiCompiler.js:188:7
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/MultiCompiler.js:270:7
    at finalCallback (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/Compiler.js:257:39)
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/Compiler.js:273:13
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:42:1)
    at AsyncSeriesHook.lazyCompileHook (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/tapable/lib/Hook.js:154:20)
    at onCompiled (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/Compiler.js:271:21)
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/Compiler.js:681:15
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (/home/prakhar/projects/techdocs_og/techdocs/website/node_modules/tapable/lib/Hook.js:154:20)
    at /home/prakhar/projects/techdocs_og/techdocs/website/node_modules/webpack/lib/Compiler.js:678:31
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ build: `docusaurus build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/prakhar/.npm/_logs/2020-12-02T16_52_03_768Z-debug.log`

mttrbrts commented on .dockerignore:
```diff
@@ -0,0 +1,22 @@
+v2-website/yarn.lock
+v2-website/node_modules
+v2-website/i18n/*
+!v2-website/i18n/en.json
```
Please refactor these to use the new folder structure.

prakharsr replied:
fixed in 4f8d81d971a4ee1142b87a86d8b1fda3c4fa8025

mttrbrts commented on .gitignore:
```diff
@@ -12,4 +12,11 @@ website/node_modules
+v2-website/yarn.lock
+v2-website/node_modules
+v2-website/i18n/*
+!v2-website/i18n/en.json
```
Please refactor these to use the new folder structure.

prakharsr replied:
fixed in 4f8d81d971a4ee1142b87a86d8b1fda3c4fa8025

mttrbrts commented on Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM node:12-slim
```
Can you upgrade to Node 14, please?

prakharsr replied:
fixed in 4f8d81d971a4ee1142b87a86d8b1fda3c4fa8025

mttrbrts commented on Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM node:12-slim
+COPY . ./
+RUN cd v2-website && npm i 
```
Should use `website` rather than `v2-website`

prakharsr replied:
fixed in d4f30ed148a9269ac3b5606992d124f7b3f66d75

mttrbrts commented on README.md:
```diff
@@ -23,12 +23,22 @@ Accord Project is an open source, non-profit, initiative working to transform co
 ## Installation
 
-To build the documentation locally:
+To build the documentation locally (docusaurus v1):
```
Now redundant and can be removed

prakharsr replied:
fixed in 4f8d81d971a4ee1142b87a86d8b1fda3c4fa8025

mttrbrts commented on README.md:
````diff
@@ -23,12 +23,22 @@ Accord Project is an open source, non-profit, initiative working to transform co
+
+To build the documentation locally (docusaurus v2):
+```
+cd ./v2-website
````
Should use `website` rather than `v2-website`

prakharsr replied:
fixed in 4f8d81d971a4ee1142b87a86d8b1fda3c4fa8025

mttrbrts commented on docs/assets/specification/ErgoSpec.Backend.ErgoBackend.jsx:
```diff
@@ -0,0 +1,250 @@
+<div>
```
How were these JSX files generated, please? Did you use the Docusarus migration scripts, for example?

Because these files are dynamically created, we should make sure that our scripts generate the correct output. This can wait until after this PR has been merged, but should be completed before merging to `master`

prakharsr replied:
According to Docusaurus v2 migration guide, they provided this website https://transform.tools/html-to-jsx for transforming html to jsx. So, I manually converted all html files to jsx.

mttrbrts replied:
Thanks, @prakhars for your diligence here, however, unfortunately, these files are stale (they don't seem to have been updated in over 2 years).`

I believe that all of the files in `docs/assets/specification` should be deleted. @jeromesimeon can you confirm, please?

mttrbrts commented on docs/ref-ergo-spec.md:
```diff
@@ -218,16 +218,16 @@ Precedence determines the order of operations in expressions with operators of d
-10 | and | logical conjunction | left to right
-11 | or | logical disjunction | left to right
-12 | < <br> > <br> <= <br> >= <br> = <br> != | less than <br> greater than <br> less or equal <br> greater or equal <br> equal <br> not equal | left to right
+--- | --- | --- | --- 1 | . <br /> ?. | field access <br /> field access of
```
The table structure has been broken here an no longer renders properly. Do you understand why?

prakharsr replied:
Yes, I used the automatic migration cli command provided by docusaurus ```npx @docusaurus/migrate migrate --mdx --page ./v1-website ./v2-website``` . Since, this is still in alpha, so it replaced the table structure incorrectly, leading to it being broken. I'll review any broken tables and similar files and fix them in the upcoming commits.

prakharsr replied:
I fixed all docs files after reviewing and comparing them in a6f0627926cce81b5b0389d9eb17b1f1ee7b4437. The error was caused because in JSX, <br> tag now needs to be closed.

mttrbrts replied:
Hi @prakharsr, no problem and thanks for the update. I'll review again this weekend.

mttrbrts reviewed (changes requested):
Incomplete review. I'll continue later

mttrbrts commented on website/scripts/build_api_md.sh:
```diff
@@ -1,32 +0,0 @@
-#!/bin/bash
```
This file and all others in `website/scripts` should be kept. They are used for generating the markdown for the API from the source code. Can you restore them, please?

prakharsr replied:
fixed in 16ef333f72f5621e906d3ef6250bafffe756b72f

mttrbrts commented on website/static/docs/accordproject-faq.md:
```diff
@@ -0,0 +1,58 @@
+---
```
What's the difference between the `./website/static/docs` directory and `./docs`, please? I assume that we only need one of them?

prakharsr replied:
We need docs in static folder because the line ```![Template-Studio-V2](/docs/assets/reference/tsv2.png)``` in ```website/versioned_docs/version-0.21/ref-cicero-ui.md``` is looking for these files in ``website/static/docs/assets/reference/tsv2.png``` thus leading to redundancy. Any way we can address this problem by changing the folder where it looks for files to direct to our docs folder ?

prakharsr replied:
Some versioned files had links like ```/docs/assets/*.png```, they were searching for these files in static folder and causing errors so earlier I copied the docs folder into static folder. Now, in adee0af8e3bbe28a853118d979b5175da5be2c1d I've fixed this issue, removing unnecessary website/docs/*  and website/static/docs/*.md while preserving website/static/docs/assets.

mttrbrts commented on website/versioned_docs/version-0.13/assets/architecture/.gitignore:
```diff
@@ -0,0 +1,9 @@
+*.out
```
It now seems that every versioned doc has a full copy of the assets folder, which I guess makes sense. However, it seems to be highly redundant, and very difficult to review. Do you have any ideas how we can make all old doc versions refer to the same set of asset files?

prakharsr replied:
fixed in 59837595ac0a3b8152da2cd975f5e2e3ebf30b66, the md files had links to images as ```(assets/*.png)```, now I've replaced all the docs and versioned docs to use the ```(/docs/assets/*.png)``` utilising the assets in ```website/static/docs/assets```. I've also removed the now unnecessary assets folders in each versioned doc thus reducing redundancy.

mttrbrts reviewed (changes requested):
@prakharsr I've added a few more comments. Mostly they are about the structure of the files and redundancy in old versions.

mttrbrts commented:
@prakharsr Adding a link to a short discussion about this PR in a request community working group call, https://vimeo.com/489140290

mttrbrts commented:
@prakharsr Also, I found the problem that is causing the build to fail...

The MDX render is failing on an empty code block in several places. To fix, you should replace `<code></code>` with `<code> </code>` (Notice the space between the elements).

There is probably a better way to fix this, and we'll need a way to do this automatically because several of the files that have this problem are auto-generated. However, this will let us preview your build on Netlify and get your PR merged so that we can start work on the other issues.

Once I fixed this issue locally, I started seeing this error:
```
Error: You are trying to create client-side redirections to paths that do not exist:
- /*
```
To fix that, I had to change the `plugins` entry in `docusaurus.config.js` to
```
  "plugins": [
    [
      "@docusaurus/plugin-client-redirects",
      {
        "fromExtensions": [
          "html"
        ]
      }
    ]
  ],
```

prakharsr commented:
@mttrbrts 
1. I fixed the build according to your instructions in d4e94acded16a3a8c7afd9d91ad4fb5d8ed83438 and it works! I setup my own netlify deploy for this project at url [https://ecstatic-lewin-144d0e.netlify.app/](url) and added the netlify deploy status badge in this project's [https://github.com/prakharsr/techdocs](url) README.md. ```[![Netlify Status](https://api.netlify.com/api/v1/badges/b6de7e93-88c3-4684-8424-298ed99b6931/deploy-status)](https://app.netlify.com/sites/ecstatic-lewin-144d0e/deploys)```. The current netlify automatically running for this project uses an old node version, that seems to be the issue why its build is failing.
2. There is a file missing ```website/static/docs/assets/advanced/late10b.png``` which is references in all versions in file ```advanced-latedelivery.md``` and in versions 0.20 & 0.21 in file ```tutorial-latedelivery.md```. As a temporary workaround, I created a duplicate of the file late10.png and renamed it as late10b.png.

mttrbrts reviewed (approved):
Congratulations @prakharsr on your first PR! The build is now working and I'm happy to merge to `docusaurus-v2`.

We value your contribution and welcome you to the Accord Project community

We'll continue to work on this branch for a few weeks to do a thorough review before we merge this into `master` and make this is default release.

Michael-Grover commented:
@jesswang-dev
